File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/projecting/PredicateMatchingProjector.java
package com.codepoetics.raffia.streaming.projecting;

import com.codepoetics.raffia.baskets.Basket;
import com.codepoetics.raffia.operations.ProjectionResult;
import com.codepoetics.raffia.operations.Projector;
import com.codepoetics.raffia.streaming.FilteringWriter;
import com.codepoetics.raffia.visitors.Visitors;
import com.codepoetics.raffia.writers.BasketWriter;

import java.math.BigDecimal;

public final class PredicateMatchingProjector<T extends BasketWriter<T>> extends StreamingProjector<T> {

  private final Projector<Basket> projector;

  public PredicateMatchingProjector(T target, FilteringWriter<T> parent, Projector<Basket> projector) {
    super(target, parent);
    this.projector =  projector;
  }

  @Override
  public FilteringWriter<T> beginObject() {
    return WeavingProjector.weavingObject(target, this, projector);
  }

  @Override
  public FilteringWriter<T> beginArray() {
    return WeavingProjector.weavingArray(target, this, projector);
  }

  @Override
  public FilteringWriter<T> end() {
    if (parent == null) {
      throw new IllegalStateException("end() called when not writing array or object");
    }

    return parent.advance(target);
  }

  @Override
  public FilteringWriter<T> key(String key) {
    throw new IllegalStateException("key() called when not writing array or object");
  }

  private FilteringWriter<T> projected(ProjectionResult<Basket> projection) {
    for (Basket basket : projection) {
      target = basket.visit(Visitors.writingTo(target));
    }
    return this;
  }

  @Override
  public FilteringWriter<T> add(String value) {
    return projected(projector.project(Basket.ofString(value)));
  }

  @Override
  public FilteringWriter<T> add(BigDecimal value) {
    return projected(projector.project(Basket.ofNumber(value)));
  }

  @Override
  public FilteringWriter<T> add(boolean value) {
    return projected(projector.project(Basket.ofBoolean(value)));
  }

  @Override
  public FilteringWriter<T> addNull() {
    return projected(projector.project(Basket.ofNull()));
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/projecting/PredicateMatchingProjector.kt
package com.codepoetics.raffia.streaming.projecting

import com.codepoetics.raffia.baskets.Basket
import com.codepoetics.raffia.operations.ProjectionResult
import com.codepoetics.raffia.java.api.Projector
import com.codepoetics.raffia.streaming.FilteringWriter
import com.codepoetics.raffia.writers.BasketWriter
import com.codepoetics.raffia.writers.writeTo

import java.math.BigDecimal

internal class PredicateMatchingProjector<T : BasketWriter<T>>(target: T, parent: FilteringWriter<T>, private val projector: Projector<Basket>) : StreamingProjector<T>(target, parent) {

    override fun beginObject(): FilteringWriter<T> {
        return WeavingProjector.weavingObject(target, this, projector)
    }

    override fun beginArray(): FilteringWriter<T> {
        return WeavingProjector.weavingArray(target, this, projector)
    }

    override fun end(): FilteringWriter<T> {
        if (parent == null) {
            throw IllegalStateException("end() called when not writing array or object")
        }

        return parent.advance(target)
    }

    override fun key(key: String): FilteringWriter<T> {
        throw IllegalStateException("key() called when not writing array or object")
    }

    private fun projected(projection: ProjectionResult<Basket>): FilteringWriter<T> {
        for (basket in projection) {
            target = basket.writeTo(target)
        }
        return this
    }

    override fun add(value: String): FilteringWriter<T> {
        return projected(projector.project(Basket.ofString(value)))
    }

    override fun add(value: BigDecimal): FilteringWriter<T> {
        return projected(projector.project(Basket.ofNumber(value)))
    }

    override fun add(value: Boolean): FilteringWriter<T> {
        return projected(projector.project(Basket.ofBoolean(value)))
    }

    override fun addNull(): FilteringWriter<T> {
        return projected(projector.project(Basket.ofNull()))
    }
}
